From 571554805ad19c10ac356583475900185969fe4a Mon Sep 17 00:00:00 2001
From: Janusz Dziedzic <janusz.dziedzic@iopsys.eu>
Date: Wed, 29 Sep 2021 08:43:17 +0200
Subject: [PATCH 07/12] wireless: fix crash when multi AP used

This is BT:
(gdb) bt
0  0x77da2070 in __syscall_ret (r=4294967294) at src/internal/syscall_ret.c:5
1  0x77de5f0c in fstatat (fd=<optimized out>, path=<optimized out>, st=0x7f298140, flag=<optimized out>) at src/stat/fstatat.c:126
2  0x00404bc5 in vlanconf_exists () at system-linux.c:264
3  0x00404bf9 in __device_get.constprop.30 (name=0x9e0060 "wlan1.sta1", create=<optimized out>, check_vlan=<optimized out>) at device.c:662
4  0x0040c8fb in get_vlan_device_chain (ifname=<optimized out>, create=<optimized out>) at vlan.c:274
5  0x00404c0b in __device_get.constprop.30 (name=0x9e0040 "wlan1.sta1", create=<optimized out>, check_vlan=<optimized out>) at device.c:668
6  0x0040c8fb in get_vlan_device_chain (ifname=<optimized out>, create=<optimized out>) at vlan.c:274
7  0x00404c0b in __device_get.constprop.30 (name=0x9e0020 "wlan1.sta1", create=<optimized out>, check_vlan=<optimized out>) at device.c:668
8  0x0040c8fb in get_vlan_device_chain (ifname=<optimized out>, create=<optimized out>) at vlan.c:274
9  0x00404c0b in __device_get.constprop.30 (name=0x9e0000 "wlan1.sta1", create=<optimized out>, check_vlan=<optimized out>) at device.c:668
...
...
149715 0x0040904f in wireless_interface_handle_link (vif=0x77cef730, ifname=0x790f74 "wlan1.sta1", up=<optimized out>) at wireless.c:330
149716 0x0040c2d9 in wireless_device_hotplug_event (add=true, name=<optimized out>) at wireless.c:1529
149717 device_hotplug_event (name=<optimized out>, add=<optimized out>) at device.c:1226
149718 0x004033c7 in handle_hotplug_msg (size=<optimized out>, data=<optimized out>) at system-linux.c:733
149719 handle_hotplug_event (u=0x42d950 <hotplug_event>, events=<optimized out>) at system-linux.c:749
149720 0x77d6719b in uloop_run_events (timeout=<optimized out>) at uloop.c:198
149721 uloop_run_timeout (timeout=<optimized out>) at uloop.c:555
149722 0x00402c9d in uloop_run () at /home/janusz/iop/ex400/s18/staging_dir/target-mipsel_24kc_musl/usr/include/libubox/uloop.h:111
149723 main (argc=<optimized out>, argv=<optimized out>) at main.c:339

Signed-off-by: Janusz Dziedzic <janusz.dziedzic@iopsys.eu>
---
 device.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/device.c b/device.c
index 50bfa8a..ff873ef 100644
--- a/device.c
+++ b/device.c
@@ -658,7 +658,7 @@ __device_get(const char *name, int create, bool check_vlan)
 
 	dev = avl_find_element(&devices, name, dev, avl);
 
-	if (!dev && strchr(name, '.')) {
+	if (!dev && strchr(name, '.') && check_vlan) {
 		if(vlanconf_exists()) {
 			// To support ethx.y.z format for 8021ad interface names
 			if (!config_is_device_present(name))
-- 
2.32.0

